### PR TITLE
Add Documenter support to generate

### PR DESCRIPTION
**Pull request by bramtayl**
This expands the default package generation to support out of the box documentation. The generic documentation is by default a formatted printout of docstrings (similar to how things work for CRAN packages).

This has huge advantages for me in terms of greatly reducing the amount of time to get a package up and running. I did feel a twinge of guilt adding Documenter to REQUIRES: if this goes through it probably should go hand in hand with stability and support over there.

The actual `makedocs` code is currently in `tests/runtests.jl`. This enables doctests to be included as part of the testing suite, and also allows doctests to count towards code coverage.

As an added bonus, I also reformatted the `generate` docstring using Markdown lists.

### CODE REVIEW

**Comment by tkelman** [comment]
wouldn't documenter be a test-only dependency based on how you're using it right now?

**Comment by bramtayl** [comment]
I'll try moving it to tests/REQUIRE and seeing what happens.

**Comment by bramtayl** [comment]
Worked like a charm, and also got tests to work1

File: src/generate.jl [hunk 1]
```diff
@@ -211,19 +214,21 @@ function tests(pkg::AbstractString; force::Bool=false,document=true,
         using Base.Test
 
         # write your own tests here
-        @test 1 == 2
+        # @test 1 == 2
```
Comment (tkelman) [inline src/generate.jl]: this is intentional

Comment (ararslan) [inline src/generate.jl]: In particular, I believe the thinking is that you shouldn't get a "free pass" by default; i.e. right out of the box the tests should fail to encourage you to write actual tests.

File: src/generate.jl [hunk 2]
```diff
@@ -302,13 +362,59 @@ function codecov(pkg::AbstractString; force::Bool=false)
...
+    pkg_name = basename(pkg)
+    genfile(pkg,"docs/make.jl",force) do io
+        print(io, """
+        using Documenter, $pkg_name
+
+        # for successful deployment, make sure to
+        # - add a gh-pages branch on github
+        # - run `import Documenter; Documenter.Travis.genkeys("$pkg_name")` in a *REPL*
+        #       and follow instructions. (Requires several unix tools. For Windows, try
+        #       adding `C:\\Program Files\\Git\\usr\\bin` to your path first
```
Comment (tkelman) [inline src/generate.jl]: this isn't a good recommendation, modifying path in any permanent way is not a good idea

File: src/generate.jl (hunk 1 above)
Comment (bramtayl) [inline src/generate.jl]: ok I can change it back. it's there to make people notice and think about tests?

File: src/generate.jl (hunk 2 above)
Comment (bramtayl) [inline src/generate.jl]: oh ok. that's how i got it to work on windows. what would be a better strategy?

**Comment by bramtayl** [comment]
Ok I tried using this to build a little package (coming shortly) and it worked! The extra steps for deployment are still a bit annoying (the travis genkeys stuff) but other than that everything should be working out of box.

File: src/generate.jl (hunk 1 above)
Comment (ararslan) [inline src/generate.jl]: Yep

**Comment by tkelman** [comment]
Doesn't Documenter have its own `generate` already? cc @mortenpi @MichaelHatherly

**Comment by bramtayl** [comment]
Cf. https://github.com/JuliaDocs/Documenter.jl/issues/144

File: src/generate.jl (hunk 2 above)
Comment (tkelman) [inline src/generate.jl]: I'd say "run from inside git-bash" would be preferable

Comment (bramtayl) [inline src/generate.jl]: oh! thanks, i didn't realize that would work.

**Comment by MichaelHatherly** [comment]
> Doesn't Documenter have its own generate already?

Yes, there's already a `Documenter.generate` available for this kind of thing. Not sure we should duplicate things here as well if possible. Improvements to Documenter's `generate` would be welcomed if there's anything here that's not already there.

Regarding running `makedocs` as part of the tests: I'd suggest avoiding that. Doctests aren't really meant for use as thorough unit tests since they're string-based rather than value based. They're mostly just for checking correctness of examples so they don't get out of date and aren't really a substitute for `@test`.

**Comment by bramtayl** [comment]
Ok :( One advantage of getting things done here in one go is that there's no need to modify existing files later (like .travis.yml, [test/]REQUIRES, README) which can be dangerous. I don't think that any of that is/should be done by `Documenter.generate`, meaning that this PR saves at least 4 steps in getting out documentation. Also, both of my existing Julia packages are tested only with doctests, pretty thoroughly IMHO (although this definitely would not work with many packages).

**Comment by bramtayl** [comment]
I saw. His issue seemed to be that doctests are "string-based rather than value based." Which could definitely make a difference in some places, but I don't think it makes a difference here.

**Comment by MichaelHatherly** [comment]
> modify existing files later (like .travis.yml, [test/]REQUIRES, README) which can be dangerous

Since the files being modified are all being tracked in a git repo I don't think it would really be dangerous to modify anything. If there's already dirty files when `Documenter.generate` is run then it can just abort and tell the user to commit those changes first.

> I saw. His issue seemed to be that doctests are "string-based rather than value based."

In addition to the string/value issue doctests won't really be able to test internal state of objects that isn't printed as part of `show` output. To test internals you then end up having to expose package internals within public documentation, generally not a good thing I'd think. Platform-specific output is also a problem (`Int64`/`Int32`, or file path separators), there's not much that can be done in Documenter to handle those in a very robust way.

**Comment by bramtayl** [comment]
> I don't think it would really be dangerous to modify anything.

Ok, in that case, I'll move this PR over to Documenter. I was going by your comment in https://github.com/JuliaDocs/Documenter.jl/issues/144

> This is kind of what `Documenter.generate` does, though without modifying files that already exist (which I wouldn't really what it to ever do).

> doctests won't really be able to test internal state of objects

I see what you mean. But I think that it's still important that testing capabilities are a strict subset of doctesting capabilities. Tests can in fact be included in doctests:

```julia
julia> Base.@test a == b;
```

though I like 

```julia
julia> a == b
true
```

much better.

**Pull request closed without merging**